mirrord/layer/src/socket.rs

```diff
@@ -379,25 +398,174 @@ impl OutgoingSelector {
                     || (select_address.port() == 0 && select_address.ip() == address.ip())
                     || select_address == address
             }
-            // TODO(alex): DNS may not trigger the filter, as the local resolved address may be
-            // different from the remote resolved.
-            AddressFilter::Name((ref name, port)) => format!("{name}:{port}")
-                .to_socket_addrs()
-                .map(|mut resolved| {
-                    resolved.any(|resolved_address| {
-                        resolved_address == SocketAddr::new(address.ip(), 0)
-                    })
-                })
-                .unwrap_or_default(),
+            // TODO(alex): We could enforce this at the type level, by converting `OutgoingWhatever`
+            // to a type that doesn't have `AddressFilter::Name`.
+            AddressFilter::Name(_) => unreachable!("BUG: We skip these in the iterator!"),
             AddressFilter::Subnet((subnet, port)) => {
                 subnet.contains(&address.ip()) && (port == 0 || port == address.port())
             }
         };
 
+        // Resolve the hostnames in the filter.
+        let resolved_hosts = match &self {
+            OutgoingSelector::Unfiltered => HashSet::default(),
+            OutgoingSelector::Remote(_) => self.resolve_dns::<true>()?,
+            OutgoingSelector::Local(_) => self.resolve_dns::<false>()?,
+        };
+        let hosts = resolved_hosts.iter();
+
+        // Return the valid address to connect, in some cases (when DNS resolving is involved),
+        // this address may be different than the one we initially received in this function.
+        Detour::Success(match self {
+            OutgoingSelector::Unfiltered => ConnectionThrough::Remote(address),
+            OutgoingSelector::Remote(list) => {
+                if !list
+                    .iter()
+                    .filter(skip_unresolved)
+                    .chain(hosts)
+                    .filter(filter_protocol)
+                    .any(any_address)
+                {
+                    // No "remote" selector matched `address`, so now we try to get the correct
+                    // address to connect to, either it's a resolved hostname, then we check our
+                    // `REMOTE_DNS_REVERSE_MAPPING` and resolve the hostname locally, or this
+                    // `address` is the one the user wants.
+                    Self::get_local_address_to_connect(address)?
+                } else {
+                    ConnectionThrough::Remote(address)
+                }
+            }
+            OutgoingSelector::Local(list) => {
+                if list
+                    .iter()
+                    .filter(skip_unresolved)
+                    .chain(hosts)
+                    .filter(filter_protocol)
+                    .any(any_address)
+                {
+                    Self::get_local_address_to_connect(address)?
```

Comment:
In what situation is this not just `Local(address)`? Please document.

Reply:
Added docs, are they clear enough?